### PR TITLE
TLS Configuration Support for QueryServer and Dispatch Client

### PR DESCRIPTION
**Pull request by anandheritage**
TLS support for the QueryServer and Dispatch client in Apache Pinot. The changes include:

TLS Configuration for QueryServer:

Addition of TLS settings to enable secure communication for the QueryServer.
Implementation of SSL context building to manage certificates and keys.
Dispatch Client Updates:

Integration of TLS support for the Dispatch client to ensure secure data dispatch.
These enhancements aim to improve the security of inter-service communication within Apache Pinot, aligning with modern security standards.

image

### CODE REVIEW

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java [hunk 1]
```diff
@@ -71,7 +88,17 @@ public void start() {
     LOGGER.info("Starting QueryServer");
     try {
       if (_server == null) {
-        _server = ServerBuilder.forPort(_port).addService(this).maxInboundMessageSize(MAX_INBOUND_MESSAGE_SIZE).build();
+        if (_tlsConfig == null) {
+          try {
+            _server =
+                ServerBuilder.forPort(_port).addService(this).maxInboundMessageSize(MAX_INBOUND_MESSAGE_SIZE).build();
+          } catch (Exception e) {
+            throw new RuntimeException("Failed to start secure QueryServer", e);
```
Comment (mcvsubbu) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java:96]: Seems like a wrong message in the exception. we are trying to start the non-ssl server here, right?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java [hunk 2]
```diff
@@ -184,4 +211,27 @@ public void cancel(Worker.CancelRequest request, StreamObserver<Worker.CancelRes
     // we always return completed even if cancel attempt fails, server will self clean up in this case.
     responseObserver.onCompleted();
   }
+
+  private SslContext buildSslContext(TlsConfig tlsConfig) {
+    LOGGER.info("Building gRPC SSL context");
+    SslContext sslContext = CLIENT_SSL_CONTEXTS_CACHE.computeIfAbsent(tlsConfig.hashCode(), tlsConfigHashCode -> {
```
Comment (mgranderath) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java:217]: Just for my own understanding; Why do we need to use a map for this? Is it simply so in the rare case that we launch multiple `QueryServer` we don't duplicate the SSLContexts?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java [hunk 3]
```diff
@@ -184,4 +211,27 @@ public void cancel(Worker.CancelRequest request, StreamObserver<Worker.CancelRes
...
+      try {
+        SSLFactory sslFactory = RenewableTlsUtils.createSSLFactoryAndEnableAutoRenewalWhenUsingFileStores(tlsConfig,
+            PinotInsecureMode::isPinotInInsecureMode);
+        SslContextBuilder sslContextBuilder = SslContextBuilder.forClient();
+        sslFactory.getKeyManagerFactory().ifPresent(sslContextBuilder::keyManager);
+        sslFactory.getTrustManagerFactory().ifPresent(sslContextBuilder::trustManager);
```
Comment (mgranderath) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java:223]: Does it make sense to extract this from here so it can easily be re-used?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java (hunk 1 above)
Comment (anandheritage) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java]: Done

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java (hunk 3 above)
Comment (anandheritage) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java:228]: The logic is refered  from GrpcQueryServer.java

**Review by mgranderath — APPROVED** [review]

**Comment by anandheritage** [comment]
@Jackie-Jiang @xiangfu0  
Can you enable the workflow here.

**Comment by anandheritage** [comment]
@xiangfu0 @npawar  Can you help me get it reviewed ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/channel/GrpcMailboxServer.java [hunk 4]
```diff
@@ -37,17 +54,33 @@
  * send by the sender of the sender/receiver pair.
  */
 public class GrpcMailboxServer extends PinotMailboxGrpc.PinotMailboxImplBase {
+  private static final Logger LOGGER = LoggerFactory.getLogger(GrpcMailboxServer.class);
```
Comment (ankitsultana) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/channel/GrpcMailboxServer.java:57]: The PR title is aimed at QueryServer. Do you also need this for the GrpcMailboxServer? If so you also need to enable this in `GrpcSendingMailbox`. If you don't need this then you can revert the changes in this file.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java [hunk 5]
```diff
@@ -58,4 +90,26 @@ public void cancel(long requestId) {
     Worker.CancelRequest cancelRequest = Worker.CancelRequest.newBuilder().setRequestId(requestId).build();
     _dispatchStub.cancel(cancelRequest, NO_OP_CANCEL_STREAM_OBSERVER);
   }
+  private SslContext buildSslContext(TlsConfig tlsConfig) {
```
Comment (ankitsultana) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java:93]: lack of new line and duplicate code. You can perhaps create a new utils class in `org.apache.pinot.common.utils.grpc`?

cc: @Jackie-Jiang if he has a better idea.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java [hunk 6]
```diff
@@ -71,7 +88,17 @@ public void start() {
     LOGGER.info("Starting QueryServer");
     try {
       if (_server == null) {
-        _server = ServerBuilder.forPort(_port).addService(this).maxInboundMessageSize(MAX_INBOUND_MESSAGE_SIZE).build();
+        if (_tlsConfig == null) {
+          try {
+            _server =
+                ServerBuilder.forPort(_port).addService(this).maxInboundMessageSize(MAX_INBOUND_MESSAGE_SIZE).build();
```
Comment (ankitsultana) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java:94]: Why is try-catch required now? Since this line is the same as before.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java [hunk 7]
```diff
@@ -36,12 +52,28 @@
...
   private final PinotQueryWorkerGrpc.PinotQueryWorkerStub _dispatchStub;
 
   public DispatchClient(String host, int port) {
-    _channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
+    this(host, port, new GrpcConfig(Collections.emptyMap()));
+  }
+  public DispatchClient(String host, int port, GrpcConfig grpcConfig) {
+    if (grpcConfig.isUsePlainText()) {
+      _channel =
+          ManagedChannelBuilder.forAddress(host, port).maxInboundMessageSize(grpcConfig.getMaxInboundMessageSizeBytes())
```
Comment (ankitsultana) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java:70]: Why are we setting max inbound message size here? I don't think it's required since dispatch client's response is usually very small

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/channel/GrpcMailboxServer.java (hunk 4 above)
Comment (anandheritage) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/channel/GrpcMailboxServer.java]: Removed the GrpcMailboxServer - If required will add as a separate PR

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java (hunk 6 above)
Comment (anandheritage) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java]: Done. Although I think try-catch should be there.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java (hunk 7 above)
Comment (anandheritage) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java]: Done.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java (hunk 5 above)
Comment (anandheritage) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java:91]: @ankitsultana can we take it as a follow-up ?
As It is used in multiple places - It will cause touching other files too.

Comment (ankitsultana) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java:91]: Hmm that should be fine. I think we should avoid duplicating code in this PR itself.

Comment (anandheritage) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java:91]: Okay - let me update the PR then

Comment (anandheritage) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java]: @ankitsultana @Jackie-Jiang 
Let me know if this looks good , or if you have any different thought for the utility.

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java [hunk 8]
```diff
@@ -41,7 +45,17 @@ class DispatchClient {
   private final PinotQueryWorkerGrpc.PinotQueryWorkerStub _dispatchStub;
 
   public DispatchClient(String host, int port) {
-    _channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
+    this(host, port, new GrpcConfig(Collections.emptyMap()));
+  }
+
+  public DispatchClient(String host, int port, GrpcConfig grpcConfig) {
```
Comment (soumitra-st) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java:51]: This constructor is only used in the other constructor to create non-tls DispatchClient. Does it mean that if gRPC TLS is enabled on the server side, then DispatchClient will always fail?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java [hunk 9]
```diff
@@ -55,16 +58,18 @@ public class QueryServer extends PinotQueryWorkerGrpc.PinotQueryWorkerImplBase {
 
   private final int _port;
   private final QueryRunner _queryRunner;
+  private final TlsConfig _tlsConfig;
```
Comment (gortiz) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java:61]: @Nullable annotation is needed here

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java [hunk 10]
```diff
@@ -55,16 +58,18 @@ public class QueryServer extends PinotQueryWorkerGrpc.PinotQueryWorkerImplBase {
...
   // query submission service is only used for plan submission for now.
   // TODO: with complex query submission logic we should allow asynchronous query submission return instead of
   //   directly return from submission response observer.
   private final ExecutorService _querySubmissionExecutorService;
 
   private Server _server = null;
 
-  public QueryServer(int port, QueryRunner queryRunner) {
+  public QueryServer(int port, QueryRunner queryRunner, TlsConfig tlsConfig) {
```
Comment (gortiz) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java:69]: @Nullabe annotation is needed here

**Comment by gortiz** [comment]
This PR is buggy. It looks like QueryServer enables TLS when it is enabled in netty config but query dispatcher is never using TLS, so multi-stage query engine ends up not working at all if TLS is enabled in netty.

**Comment by anandheritage** [comment]
Let me know if you want me to revert this for now and maybe we discuss if major changes are required ?
